File: Lab7/leetcode/TastScheduler/solution.cpp
class Solution
{
public:
  int leastInterval(vector<char> &tasks, int n)
  {
    unordered_map<char, int> taskCount;
    for (char task : tasks)
    {
      taskCount[task]++;
    }

    priority_queue<int> maxHeap;
    for (const auto &entry : taskCount)
    {
      maxHeap.push(entry.second);
    }

    int totalIntervals = 0;
    queue<pair<int, int>> cooldownQueue;
    while (!maxHeap.empty() || !cooldownQueue.empty())
    {
      totalIntervals++;
      if (!maxHeap.empty())
      {
        int currentTaskCount = maxHeap.top();
        maxHeap.pop();
        if (currentTaskCount > 1)
        {
          cooldownQueue.push({currentTaskCount - 1, totalIntervals + n});
        }
      }

      if (!cooldownQueue.empty() && cooldownQueue.front().second == totalIntervals)
      {
        maxHeap.push(cooldownQueue.front().first);
        cooldownQueue.pop();
      }
    }

    return totalIntervals;
  }
};

File: Lab7/leetcode/TastScheduler/solution.java
package Lab7.leetcode.TastScheduler;

public class Solution {
  public int leastInterval(char[] tasks, int n) {
    Map<Character, Integer> taskCount = new HashMap<>();
    for (char task : tasks) {
      taskCount.put(task, taskCount.getOrDefault(task, 0) + 1);
    }

    PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
    maxHeap.addAll(taskCount.values());

    int totalIntervals = 0;
    Queue<int[]> cooldownQueue = new LinkedList<>();

    while (!maxHeap.isEmpty() || !cooldownQueue.isEmpty()) {
      totalIntervals++;

      if (!maxHeap.isEmpty()) {
        int currentTaskCount = maxHeap.poll();
        if (currentTaskCount > 1) {
          cooldownQueue.offer(new int[] { currentTaskCount - 1, totalIntervals + n });
        }
      }

      if (!cooldownQueue.isEmpty() && cooldownQueue.peek()[1] == totalIntervals) {
        maxHeap.offer(cooldownQueue.poll()[0]);
      }
    }

    return totalIntervals;
  }
}
